R/as-data-frame.R

```diff
@@ -1,34 +1,62 @@
-CrunchDataFrame <- function (dataset) {
+CrunchDataFrame <- function (dataset, selection = NULL) {
     ## S3 constructor method for CrunchDataFrame. terms.formula doesn't seem
     ## to like S4 subclass of environment
     stopifnot(is.dataset(dataset))
     out <- new.env()
     out$.crunchDataset <- dataset
+    out$.names <- aliases(allVariables(dataset))
+    # set the order of the dataset based on selection, if no selection is 
+    # given return all rows in the dataset in the order they appear
+    out$.order <- selection
+    
     with(out, {
         ## Note the difference from as.environment: wrapped in as.vector
-        for (a in aliases(allVariables(dataset))) {
-            eval(substitute(delayedAssign(v, as.vector(.crunchDataset[[v]])),
-                list(v=a)))
+        for (.a in out$.names) {
+            eval(substitute(delayedAssign(v, {
+                if (is.null(.order)) {
+                    as.vector(.crunchDataset[[v]])
+                } else {
+                    # The API response for selected rows is in the order of the
+                    # dataset, so we need to reorder it before we return it. 
+                    # `match(.order, sort(unique(.order)))`` gets the dense 
+                    # ranking of the ordering so that the non-duplicated 
+                    # results returned in crunch order can be used in the order
+                    # specified in .order.
+                    as.vector(.crunchDataset[[v]][.order])[match(.order, sort(unique(.order)))]
+                }
+                }),
+                list(v=.a)))
         }
     })
+    
     class(out) <- "CrunchDataFrame"
     return(out)
 }
 
 setOldClass("CrunchDataFrame")
+setClass(Class = "CrunchDataFrame", contains = "data.frame")
```

Comment:
Sure, it comes from `sp` here is the code for `SpatialPolygonsDataFrame` https://github.com/edzer/sp/blob/master/R/Class-SpatialPolygonsDataFrame.R there are numerous others as well: https://github.com/edzer/sp/tree/master/R We likely will only use polygons, but there's no reason we might not in the future use points or other shapes, so I'm trying to keep that option open.